packages/puppeteer-core/src/api/Page.ts

```diff
@@ -1306,6 +1307,91 @@ export abstract class Page extends EventEmitter<PageEvents> {
+    //        domain string is a %x2E (".") character.
+    //     *  The string is a host name (i.e., not an IP address).
+    // https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.3
+    const cookieDomain = cookie.domain.toLowerCase();
+    const urlHostname = parsedUrl.hostname.toLowerCase();
+    if (cookieDomain === urlHostname) {
+      // The domain string and the string are identical.
+      return true;
+    }
+    if (net.isIP(urlHostname)) {
+      // The string should be a host name (i.e., not an IP address).
+      return false;
+    }
+    if (!urlHostname.endsWith(cookieDomain)) {
+      // The domain string is a suffix of the string.
+      return false;
+    }
+    // The last character of the string that is not included in the domain string is a
+    // %x2E (".") character.
+    return urlHostname.endsWith("." + cookieDomain);
+  }
+
+  static #testCookieUrlMatchPath(cookie: Protocol.Network.Cookie, parsedUrl: URL): boolean {
+    // Check paths match according to the spec:
+    //   A request-path path-matches a given cookie-path if at least one of
+    //   the following conditions holds:
+    //   o  The cookie-path and the request-path are identical.
+    //   o  The cookie-path is a prefix of the request-path, and the last
+    //      character of the cookie-path is %x2F ("/").
+    //   o  The cookie-path is a prefix of the request-path, and the first
+    //      character of the request-path that is not included in the cookie-
+    //      path is a %x2F ("/") character.
+    // https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.4
+    const uriPath = parsedUrl.pathname;
+    const cookiePath = cookie.path;
+
+    console.log("!!@@##", uriPath, cookiePath);
+
+    if (uriPath === cookiePath) {
+      // The cookie-path and the request-path are identical.
+      return true;
+    }
+    // * The cookie-path is a prefix of the request-path, and the last character of the
+    //   cookie-path is %x2F ("/").
+    // * The cookie-path is a prefix of the request-path, and the first character of the
+    //   request-path that is not included in the cookie-path is a %x2F ("/") character.
+    if (uriPath.startsWith(cookiePath)) {
+      // The cookie-path is a prefix of the request-path.
+      if (cookiePath.endsWith('/')) {
+        // The last character of the cookie-path is %x2F ("/").
+        return true;
+      }
+      if (uriPath[cookiePath.length] === '/') {
+        // The first character of the request-path that is not included in the cookie-path
+        // is a %x2F ("/") character.
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Checks the cookie matches the URL according to the spec:
+   * * https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.3
+   * * https://datatracker.ietf.org/doc/html/rfc6265#section-5.1.4
+   */
+  static testCookieUrlMatch(cookie: Protocol.Network.Cookie, url: string): boolean {
```

Comment:
Not sure what is the best place for this methods, so I decided to keep it here for now.